File: 30-days-of-code/day03/solution.ex
defmodule Solution do
  require Integer

  def run do
    parse_input()
      |> judge_number
      |> print_results
  end

  defp parse_input do
    {value, _} = IO.read(:line) |> Integer.parse
    value
  end

  defp judge_number(n) when Integer.is_odd(n) do
    "Weird"
  end

  defp judge_number(n) do
    cond do
      2 <= n && n <= 5 ->
        "Not Weird"
      6 <= n && n <= 20 ->
        "Weird"
      true ->
        "Not Weird"
    end
  end

  defp print_results(judgment) do
    IO.write judgment
  end

end

Solution.run

File: 30-days-of-code/day03/solution.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

func main() {
	n := parseInput(os.Stdin)
	judgment := judgeNumber(n)
	printResults(os.Stdout, judgment)
}

// --------------------------------------------------------------------
// Computation

func judgeNumber(n uint64) string {
	switch {
	case isOdd(n):
		return "Weird"
	case n >= 2 && n <= 5:
		return "Not Weird"
	case n >= 6 && n <= 20:
		return "Weird"
	default:
		return "Not Weird"
	}
}

func isOdd(n uint64) bool {
	return n % 2 != 0
}

// --------------------------------------------------------------------
// Input/Output

func parseInput(input io.Reader) uint64 {
	scanner := bufio.NewScanner(input)
	return parseUint(scanner, "number")
}

func printResults(out io.Writer, judgment string) {
	fmt.Println(judgment)
}

// --------------------------------------------------------------------
// Framework

type parser func(input string) (interface{}, error)

func getNextInput(scanner *bufio.Scanner, parseFn parser, description string) interface{} {
	scanner.Scan()
	token := scanner.Text()
	value, err := parseFn(token)
	if err != nil {
		panic(fmt.Sprintf("invalid %s: %v", description, token))
	}
	return value
}

func parseUint(scanner *bufio.Scanner, description string) uint64 {
	parseFn := func(s string) (interface{}, error) {
		return strconv.ParseUint(s, 10, 64)
	}

	value := getNextInput(scanner, parseFn, description)
	return value.(uint64)
}
